src/OpenApi/src/Services/Schemas/OpenApiSchemaService.cs

```diff
@@ -457,5 +457,138 @@ private async Task InnerApplySchemaTransformersAsync(IOpenApiSchema inputSchema,
+                {
+                    return node;
+                }
+
+                visitedRefs.Add(refString);
+
+                try
+                {
+                    // Resolve the reference path to the actual schema content
+                    // to avoid relative references
+                    var resolvedNode = ResolveReference(refString, rootSchema);
+                    if (resolvedNode != null)
+                    {
+                        return resolvedNode.DeepClone();
+                    }
+                }
+                catch (InvalidOperationException)
+                {
+                    // If resolution fails due to invalid path, return the original reference
+                    // This maintains backward compatibility while preventing crashes
+                }
+                finally
+                {
+                    // Remove from visited set to allow the same reference in different branches
+                    visitedRefs.Remove(refString);
+                }
+
+                // If resolution fails, return the original reference
+                return node;
+            }
+
+            // Process all properties recursively
+            var newObject = new JsonObject();
+            foreach (var property in jsonObject)
+            {
+                if (property.Value != null)
+                {
+                    var processedValue = ResolveReferencesRecursive(property.Value, rootSchema, visitedRefs);
+                    newObject[property.Key] = processedValue?.DeepClone();
+                }
+                else
+                {
+                    newObject[property.Key] = null;
+                }
+            }
+            return newObject;
+        }
+        else if (node is JsonArray jsonArray)
+        {
+            var newArray = new JsonArray();
+            for (var i = 0; i < jsonArray.Count; i++)
+            {
+                if (jsonArray[i] != null)
+                {
+                    var processedValue = ResolveReferencesRecursive(jsonArray[i]!, rootSchema, visitedRefs);
+                    newArray.Add(processedValue?.DeepClone());
+                }
+                else
+                {
+                    newArray.Add(null);
+                }
+            }
+            return newArray;
+        }
+
+        // Return non-$ref nodes as-is
+        return node;
+    }
+
+    private static JsonNode? ResolveReference(string refPath, JsonNode rootSchema)
+    {
+        if (string.IsNullOrWhiteSpace(refPath))
+        {
+            throw new InvalidOperationException("Reference path cannot be null or empty.");
+        }
+
+        if (!refPath.StartsWith(OpenApiConstants.RefPrefix, StringComparison.Ordinal))
+        {
+            throw new InvalidOperationException($"Only fragment references (starting with '{OpenApiConstants.RefPrefix}') are supported. Found: {refPath}");
+        }
+
+        var path = refPath.TrimStart('#').TrimStart('/');
```

Comment:
Could use the [`TrimStart(params ReadOnlySpan<char>)`](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/String.Manipulation.cs,2410) overload?

```suggestion
        var path = refPath.TrimStart('#', '/');
```